McEngine/src/GUI/CBaseUICanvas.cpp

```diff
@@ -0,0 +1,258 @@
+
+void CBaseUICanvas::insertSlotBack(Slot *slot, Slot *index){
+	if (slot == NULL || index == NULL) return;
+
+	for (int i=0; i<m_vSlots.size(); i++){
+		if (m_vSlots[i] == index){
+			m_vSlots.insert(m_vSlots.begin() + clamp<int>(i+1, 0, m_vSlots.size()), slot);
+			return;
+		}
+	}
+
+	debugLog("Warning: CBaseUICanvas::insertSlotBack() couldn't find index\n");
+}
+
+void CBaseUICanvas::insertSlotBack(CBaseUIElement *element, Slot *index){
+	if (element == NULL || index == NULL) return;
+
+	Slot *newSlot = new Slot;
+	newSlot->element = element;
+
+	for (int i=0; i<m_vSlots.size(); i++){
+		if (m_vSlots[i] == index){
+			m_vSlots.insert(m_vSlots.begin() + clamp<int>(i+1, 0, m_vSlots.size()), newSlot);
+			return;
+		}
+	}
+
+	debugLog("Warning: CBaseUICanvas::insertSlotBack() couldn't find index\n");
+}
+
+void CBaseUICanvas::insertSlotBack(CBaseUIElement *element, bool scaleByHeightOnly, Slot *index){
+	if (element == NULL || index == NULL) return;
+
+	Slot *newSlot = new Slot;
+	newSlot->element = element;
+	newSlot->scaleByHeightOnly = scaleByHeightOnly;
+
+	for (int i=0; i<m_vSlots.size(); i++){
+		if (m_vSlots[i] == index){
+			m_vSlots.insert(m_vSlots.begin() + clamp<int>(i+1, 0, m_vSlots.size()), newSlot);
+			return;
+		}
+	}
+
+	debugLog("Warning: CBaseUICanvas::insertSlotBack() couldn't find index\n");
+}
+
+void CBaseUICanvas::removeSlot(Slot *slot){
+	for (int i=0; i<m_vSlots.size(); i++)
+	{
+		if (m_vSlots[i] == slot){
+			m_vSlots.erase(m_vSlots.begin()+i);
+			return;
+		}
+	}
+
+	debugLog("Warning: CBaseUICanvas::removeSlot() couldn't find element\n");
+}
+
+void CBaseUICanvas::deleteSlot(Slot *slot){
+	for (int i=0; i<m_vSlots.size(); i++){
+		if (m_vSlots[i] == slot){
+			delete m_vSlots[i]->element;
+			delete m_vSlots[i];
+			m_vSlots.erase(m_vSlots.begin()+i);
+			return;
+		}
+	}
+
+	debugLog("Warning: CBaseUICanvas::removeSlot() couldn't find element\n");
+}
+
+Slot *CBaseUICanvas::getSlotByElementName(UString name){
+	for (int i=0; i<m_vSlots.size(); i++){
+		if (m_vSlots[i]->element->getName() == name)
+			return m_vSlots[i];
+	}
+
+	debugLog("Error: CBaseUICanvas::getSlotByElementName() \"%s\" does not exist!!!\n", name.toUtf8());
+	return NULL;
+}
+
+std::vector<CBaseUIElement*> CBaseUICanvas::getAllElements(){
+	std::vector<CBaseUIElement*> elements;
+
+	for (int i=0; i<m_vSlots.size(); i++){
+		elements.push_back(m_vSlots[i]->element);
+	}
+
+	return elements;
+}
+
+void CBaseUICanvas::draw(Graphics *g){
+	if (!m_bVisible) return;
+
+	Vector2 posNormal;
+	Vector2 sizeNormal;
+	for (int i=0; i<m_vSlots.size(); i++){
+		sizeNormal = m_vSlots[i]->element->getSize();
+
+		if (!m_vSlots[i]->scaleByHeightOnly){
+			m_vSlots[i]->element->setSizeAbsolute(sizeNormal * m_vSize);
+		}
+
+		else{
+			m_vSlots[i]->element->setSizeAbsolute(sizeNormal.x * ((m_vSize.y / 9) * 16), sizeNormal.y * m_vSize.y);
+		}
+
+		posNormal = m_vSlots[i]->element->getPos();
+		m_vSlots[i]->element->setPosAbsolute(m_vPos + posNormal * m_vSize);

```

Comment:
You really shouldn't call anything besides other draw() functions within a draw().
If the position and size have to be updated, then it should happen during update().
(This could have a cascading effect of events that would get handled _every frame_ by lots of elements, which would decrease the amount of optimization possible for the low level renderer in the future. It kind of defeats the purpose of having separate update() and draw() functions)